File: src/12-challenge/solution.ts
type Ornament = '#' | '*' | '@' | '^' | 'o'

export default function calculatePrice(ornaments: string): number | undefined {
	const prices: Record<Ornament, number> = {
		'#': 50,
		'*': 1,
		'@': 100,
		'^': 10,
		'o': 5,
	}

	if (ornaments.match(/[^#*@^o]/)) return undefined

	let totalPrice: number = prices[ornaments.at(-1) as Ornament]

	for (let i = 0; i < ornaments.length - 1; i++) {
		const currentOrnament: number = prices[ornaments[i] as Ornament]
		const nextOrnament: number = prices[ornaments[i + 1] as Ornament]

		totalPrice += currentOrnament - nextOrnament < 0 ? -currentOrnament : currentOrnament
	}

	return totalPrice
}

File: src/12-challenge/solution.py
import re


def calculate_price(ornaments: str) -> int | None:
    prices: dict[str, int] = {
        "#": 50,
        "*": 1,
        "@": 100,
        "^": 10,
        "o": 5,
    }

    if re.search(pattern=r"[^#*@^o]", string=ornaments):
        return None

    total_price: int = prices[ornaments[-1]]

    for i in range(len(ornaments) - 1):
        current_ornament: int = prices[ornaments[i]]
        next_ornament: int = prices[ornaments[i + 1]]

        total_price += (
            -current_ornament
            if current_ornament - next_ornament < 0
            else current_ornament
        )

    return total_price
